CASMAUTO 32 : adding IUF BATS

Pankhuri-Rajesh opened the pull request:
## Summary and Scope

Adding the files for IUF test cases which have been automated and are to be made a part of CSM 1.6 
Changes-
          1. All the goss files for IUF placed under **ncn** folder. (file name starts with "goss-iuf-")
          2. All python scripts are placed under src/csm-testing/tests and lib.
          3. iuf_run_setup is the folder containing the dummy-product tar file which will be used for running the tests (mainly IUF stages till prepare-images) and the manifest files which is placed in goss-testing/scripts.
          4. New variables-iuf.yaml file added under vars folder.

## Issues and Related PRs

* [CASMAUTO-46](https://jira-pro.it.hpe.com:8443/browse/CASMAUTO-46)
* [CASMAUTO-32](https://jira-pro.it.hpe.com:8443/browse/CASMAUTO-32)

## Testing

Scripts have been tested by directly installing the modules on the system. Testing with virtual environment is in progress.

### Tested on:

  * starlord

## Pull Request Checklist

- [ ] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [x] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

mharding-hpe commented on requirements.txt:
```diff
@@ -6,7 +6,11 @@
 colorama
 cray-product-catalog
 PyYAML
+requests==2.25.1
```
The `requests` package is already available in the virtual environment from the system Python packages. See later in this file.
```suggestion
```
Is there a reason you need this specific version of the package?

mharding-hpe commented on requirements.txt:
```diff
@@ -6,7 +6,11 @@
 PyYAML
+requests==2.25.1
 requests-retry-session
+kubernetes==26.1.0
```
The `kubernetes` package is already available in the virtual environment from the system Python packages. See later in this file. 
```suggestion
```

Is there a reason you need this specific version of the package?

mharding-hpe commented on requirements.txt:
```diff
@@ -6,7 +6,11 @@
 requests-retry-session
+kubernetes==26.1.0
+python-keycloak==0.27.1
+urllib3==1.26.20
```
The `urllib3` package is already available in the virtual environment from the system Python packages. See later in this file.
```suggestion
```
Is there a reason you need this specific version of the package?

Pankhuri-Rajesh replied:
yes the keycloak version needs this version of urllib3

mharding-hpe replied:
That doesn't seem to be the case from my testing -- are you sure?

mharding-hpe commented on requirements.txt:
```diff
@@ -6,7 +6,11 @@
+requests==2.25.1
 requests-retry-session
+kubernetes==26.1.0
+python-keycloak==0.27.1
```
In the requirements.txt file, we generally do not pin the version. The only times we should specify versions in this file is if they are true requirements (as in,. the code won't work without a certain minimum or maximum package version).

For cases where we just want to pin the version that we are choosing to use, that is what the constraints.txt file is for.

```suggestion
python-keycloak
```

mharding-hpe commented on constraints.txt:
```diff
@@ -5,3 +5,6 @@ PyYAML>=6.0.1,<6.1
 requests-retry-session>=0.1,<0.2
+kubernetes==26.1.0
+python-keycloak==0.27.1
+urllib3==1.26.20
```
For the packages we are using from the system Python environment, if we are going to specify a version here, it should match the version installed in the CSM 1.6 images.
```suggestion
urllib3==1.25.10
```

Is there a reason that version will not work with your tests? If so, we'll need to stop using the system version of that package.

mharding-hpe commented on goss-testing/scripts/check_iuf_abort.sh:
This file needs a copyright header, like the other shell scripts in that directory.

mharding-hpe commented on goss-testing/scripts/python/check_workflow_template.py:
All New Python code should be added under `src/csm_testing`.
New Python test code should be added under `src/csm_testing/tests`.
New Python test utility code should be added under `src/csm_testing/tools`.
New Python general library code should be added under `src/csm_testing/lib`.
See the details in the repo README.

Pankhuri-Rajesh replied:
all the files have been shifted to appropriate folders. they have also been tested on starlord.

mharding-hpe commented on goss-testing/tests/iuf/goss-iuf-activity-abort.yaml:
Unless I am mistaken, I don't think any of the things under `tests/iuf` are going to be installed by the RPM. 

All Goss test YAML files should be under one of the following directories:
* `goss-testing/tests/ncn` - For tests that are run on NCNs but NOT the PIT node
* `goss-testing/tests/livecd` - For tests that are run on the PIT node but NOT regular NCNs
* `goss-testing/tests/common` - For tests that are run on both regular NCNs and the PIT node

The new tests should be categorized in this way also. You can name the files with the 'goss-iuf-' prefix within those folders to indicate that they are related to IUF.

Pankhuri-Rajesh replied:
we will make the changes and commit again

jacobsalmela reviewed (commented):
Will any tests need to be skipped on Vshasta?  If so, a key should be added for that.

Pankhuri-Rajesh commented:
> Will any tests need to be skipped on Vshasta? If so, a key should be added for that.

all the tests can run on Vshashta

mharding-hpe commented on constraints.txt:
```diff
@@ -5,3 +5,6 @@ PyYAML>=6.0.1,<6.1
 requests-retry-session>=0.1,<0.2
+kubernetes==26.1.0
+requests-oauthlib==0.8.0
+urllib3==1.26.20
```
If you are going to specify this here, we want it to match the version installed on the CSM 1.6.0 images.

```suggestion
urllib3==1.25.10
```

mharding-hpe commented on goss-testing/scripts/check_iuf_abort.sh:
```diff
@@ -0,0 +1,28 @@
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+
+python3 /opt/cray/tests/install/ncn/scripts/python/iuf_run /opt/cray/tests/install/ncn/scripts/iuf_run_setup test-activity & \
```
You should call the script directly, to ensure that it uses the Python binary in the virtual environment.

```suggestion
/opt/cray/tests/install/ncn/scripts/python/iuf_run /opt/cray/tests/install/ncn/scripts/iuf_run_setup test-activity & \
```

mharding-hpe commented on goss-testing/scripts/iuf_run_setup/management-bootprep.yaml:
```diff
@@ -0,0 +1,138 @@
+# (C) Copyright 2022-2023 Hewlett Packard Enterprise Development LP
```
```suggestion
# (C) Copyright 2024 Hewlett Packard Enterprise Development LP
```

mharding-hpe commented on goss-testing/scripts/workflows/iuf_base_workflow.yaml:
File needs a copyright/license header

mharding-hpe commented on goss-testing/scripts/workflows/update_cpc_workflow.yaml:
File needs a copyright/license header

mharding-hpe commented on goss-testing/scripts/workflowtemplates/iuf_base_template.yaml:
File needs a copyright/license header

mharding-hpe commented on goss-testing/scripts/workflowtemplates/update_product_catalog_template.yaml:
File needs a copyright/license header

mharding-hpe commented on goss-testing/tests/ncn/README.md:
This file does not make sense to put in the `ncn` directory, since it does not apply to a lot of the contents of that directory.

Pankhuri-Rajesh replied:
true, removing it

mharding-hpe commented on goss-testing/vars/variables-iuf.yaml:
```diff
@@ -0,0 +1,50 @@
+#
+# MIT License
+#
+# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
```
```suggestion
# (C) Copyright 2024 Hewlett Packard Enterprise Development LP
```

mharding-hpe commented on goss-testing/tests/ncn/goss-validate-compute-uan-bootprep.yaml:
```diff
@@ -0,0 +1,43 @@
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+
+command:
+  {{ $testlabel := "iuf_validate_compute_uan_bootprep" }}
+  {{$testlabel}}:
+    title: iuf validate compute and uan bootprep
```
```suggestion
  {{ $testlabel := "sat_validate_compute_uan_bootprep" }}
  {{$testlabel}}:
    title: SAT validate compute and UAN bootprep
```

mharding-hpe commented on goss-testing/tests/ncn/goss-validate-compute-uan-bootprep.yaml:
You may want to rename this file to have the `goss-sat-` prefix, to be consistent with how you named the IUF test files.

mharding-hpe commented on goss-testing/tests/ncn/goss-validate-management-bootprep.yaml:
```diff
@@ -0,0 +1,43 @@
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+
+command:
+  {{ $testlabel := "iuf_validate_management_bootprep" }}
+  {{$testlabel}}:
+    title: iuf validate management bootprep
```
```suggestion
  {{ $testlabel := "sat_validate_management_bootprep" }}
  {{$testlabel}}:
    title: SAT validate management bootprep
```

mharding-hpe commented on goss-testing/tests/ncn/goss-validate-management-bootprep.yaml:
You may want to rename this file to have the `goss-sat-` prefix, to be consistent with how you named the IUF test files.

mharding-hpe commented on goss-testing/tests/ncn/goss-validate-iuf-manifest.yaml:
This file should be renamed to have the prefix `goss-iuf-` to be consistent with the other IUF tests.

mharding-hpe commented on goss-testing/tests/ncn/goss-validate-iuf-manifest.yaml:
```diff
@@ -0,0 +1,46 @@
+    exec: |-
+
+      # validate iuf product manifest
+      echo "Test case:validate_iuf_manifest"
```
Why do you include this echo line? How are you running this Goss test where you are able to see that output?

mharding-hpe commented on requirements.txt:
```diff
@@ -7,6 +7,9 @@ colorama
 requests-retry-session
+python-keycloak
+requests-oauthlib
+urllib3
```
This will already be available, from the system. See below in the file.
```suggestion
```

mharding-hpe commented on requirements.txt:
```diff
@@ -7,6 +7,9 @@ colorama
 PyYAML
 requests-retry-session
+python-keycloak
+requests-oauthlib
```
Since this is so closely tied to the requests package, we should instead specify that we'll use the system copy, later in the file.
```suggestion
```

Below the line `#system_python:requests`, just add the line `#system_python:requests-oauthlib`

mharding-hpe reviewed (changes requested):
Have you installed the RPMs on a system and tested to make sure your additions work? I am seeing indications that this probably is not the case. Please do this before requesting a re-review.

mharding-hpe commented on src/csm_testing/tests/iuf_apis/__main__.py:
This Python file is almost one thousand lines long. It should be separated into multiple files. There is no requirement that each test put all of its code in the `__main__.py` file. That file is just the entry point to the test. It is allowed (and encouraged) to split the code up logically into other files. It makes it much easier to understand and debug.

mharding-hpe commented on src/csm_testing/tests/iuf_apis/__main__.py:
```diff
@@ -0,0 +1,799 @@
+        sys.exit(1)
+
+def main():
+    try:
+        os.makedirs(MEDIA_DIR, exist_ok=True)
+        print(f"INFO: Directory {MEDIA_DIR} created successfully.")
+    except OSError as e:
+        print(f"ERROR: Failed creating directory {MEDIA_DIR}: {e}")
+        sys.exit(1)
+    
+    try:
+        shutil.copy(f"{tar_dir}/{FOLDER_NAME}.tar.gz", MEDIA_DIR)
+        shutil.copy(f"{tar_dir}/product_vars.yaml", MEDIA_DIR)
+        shutil.copy(f"{tar_dir}/management-bootprep.yaml", MEDIA_DIR)
+        print(f"INFO: Files copied successfully to {MEDIA_DIR}.")
+    except IOError as e:
+        print(f"ERROR: Failed copying files: {e}")
+        sys.exit(1)
```
This code is almost identical to lines 45-60 in `iuf_run/__main__.py`. Unless there is a good reason not to, common code should be moved into shared modules under `src/csm_testing/lib`, and then imported by the tests.

mharding-hpe commented on src/csm_testing/tests/validate_iuf_manifest/__main__.py:
```diff
@@ -0,0 +1,99 @@
+    """Custom exception for product manifest validation errors."""
+    pass
+
+def load_yaml(file_path):
```
This same function is also defined in the `sat_validate_bootprep` test, with the primary difference being the names of the Exceptions being raised. This is also a good candidate to move into a shared module. That module can define and raise its own Exception, which the tests can each handle as appropriate.

mharding-hpe commented on src/csm_testing/tests/iuf_stages_cleanup/__main__.py:
```diff
@@ -0,0 +1,187 @@
+from kubernetes import client, config
+import sys
+
+IMAGE = "registry.local/artifactory.algol60.net/csm-docker/stable/product-deletion-utility:1.0.1"
+NAMESPACE = "services"
+CONFIGMAP_NAME = "cray-product-catalog"
+PRODUCT_NAME = "dummy"
+PRODUCT_VERSION = "1.0.0"
+DELETION_FILE_PATH="/etc/cray/upgrade/csm/iuf/deletion"
+NEXUS_URL = "https://packages.local"
+REPO_URL = f"{NEXUS_URL}/service/rest/v1/repositories/yum/hosted"
```
Most of these fields are also defined in the `iuf_stages_cleanup` test. I am guessing that if you change them in one test, you'd also need to change them in the other test. That means they should probably be moved into a separate module that is shared and imported by both tests.

mharding-hpe commented on src/csm_testing/tests/iuf_apis/__main__.py:
```diff
@@ -0,0 +1,799 @@
+import base64
+from keycloak import KeycloakOpenID
+import urllib3
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
```
This line probably should not be hidden in the midst of the import statements. :)

mharding-hpe commented on src/csm_testing/tests/sat_validate_bootprep/__main__.py:
```diff
@@ -0,0 +1,95 @@
+"""
+
+import yaml
+from jsonschema import validate, ValidationError, SchemaError
```
This test imports from `jsonschema`, which is not part of the standard Python library. Therefore it needs to be added to the `requirements.txt` and `constraints.txt` files.

mharding-hpe commented on src/csm_testing/tests/iuf_stages_cleanup/__init__.py:
```diff
@@ -0,0 +1 @@
+from .__main__ import get_nexus_credentials, get_ca_certificates
```
I don't see any other code trying to import from this module. Why is this line needed?

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,438 @@
+from csm_testing.tests.iuf_stages_cleanup import get_nexus_credentials, get_ca_certificates
+
+MEDIA_DIR = "/etc/cray/upgrade/csm/automation-tests"
+PRODUCT_CATALOG_CONFIG_MAP_NAMESPACE='services'
```
You should not put code in the middle of the import statements. Unless there is an important reason why you cannot, you should put all of the import statements first, and only then start variable and function declarations.

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,438 @@
+MEDIA_DIR = "/etc/cray/upgrade/csm/automation-tests"
+PRODUCT_CATALOG_CONFIG_MAP_NAMESPACE='services'
+
+from cray_product_catalog.query import *
```
Doing `import *` is bad practice in Python unless there is a compelling reason why it is necessary. Instead, just import the module, and reference the things you need by name. Or just import the exact things you need.

mharding-hpe commented on src/csm_testing/tests/iuf_pre_checks/__main__.py:
```diff
@@ -0,0 +1,348 @@
+import sys
+import requests
+import shutil
+from packaging import version
```
`packaging` is not part of the Python standard library. This needs to be added to `requirements.txt` and `constraints.txt`

mharding-hpe commented on constraints.txt:
Need to list `python-keycloak` in this file

mharding-hpe commented:
[CASMAUTO-32](https://jira-pro.it.hpe.com:8443/browse/CASMAUTO-32) is marked as done. Is that the correct ticket for this work?

Pankhuri-Rajesh commented:
> [CASMAUTO-32](https://jira-pro.it.hpe.com:8443/browse/CASMAUTO-32) is marked as done. Is that the correct ticket for this work?

[CASMAUTO-46](https://jira-pro.it.hpe.com:8443/browse/CASMAUTO-46) is the ticket being used to track this PR

mharding-hpe commented on goss-testing/scripts/check_iuf_abort.sh:
```diff
@@ -0,0 +1,28 @@
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+
+/opt/cray/tests/install/ncn/scripts/python/iuf_run /opt/cray/tests/install/ncn/scripts/iuf_run_setup test-activity & \
```
I assume that is meant to be `&&` not `&`
```suggestion
/opt/cray/tests/install/ncn/scripts/python/iuf_run /opt/cray/tests/install/ncn/scripts/iuf_run_setup test-activity && \
```

mharding-hpe reviewed (changes requested):
Looking through the Python code, I can see several packages being imported that are not included in the requirements.txt file:
jsonschema
packaging
python-keycloak

mharding-hpe commented:
We also recently added pylint to the build pipeline, and it looks like that is failing when building this PR branch.

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
All Python source code files need the standard copyright/license headers

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+        return self._token
+
+
+class ApiInterface_no_token(object):
```
Almost all of the content of this class is duplicated in the ApiInterface class. This should probably be changed into a base class, and then have the ApiInterface class as a subclass. That would avoid so much code duplication (and reduces the chance that we find a bug and forget to fix it in both places).

Srinivas-Anand-HPE replied:
will modify and inherit ApiInterface from ApiInterface_no_token Class but requests method is different as token has None value in ApiInterface_no_token so over writing it

Srinivas-Anand-HPE replied:
will remove ApiInterface_no_token class and modify ApiInterface class to be used directly by making token as parameter

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+        return self._token
+
+
+class ApiInterface_no_token(object):
```
In Python 3, there is no need to specify that a class is a subclass of `object`. It is implied.

```suggestion
class ApiInterface_no_token:
```

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+
+
+class ApiInterface_no_token(object):
+    def __init__(self, apiurl="https://api-gw-service-nmn.local/apis", resource="/iuf/v1"):
```
We should specify type annotations as much as possible. It makes it much easier to understand code when we specify the allowed types for parameters.
```suggestion
    def __init__(self, apiurl: str="https://api-gw-service-nmn.local/apis", resource: str="/iuf/v1"):
```

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+        url=self.apiurl + self.resource + path
+
+        headers = dict()
+        try:
```
What is the purpose of this try block? I don't see anything in the following two lines that would raise an exception. Which one are you concerned about?

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+
+        headers = dict()
+        try:
+            token = None
+            headers["Authorization"] = f"Bearer {token}"
```
Would we not want to just omit the Authorization header altogether, rather than use the None value for the token? Is there a reason to include it?

Srinivas-Anand-HPE replied:
yes , It is to get 403 Forbiden error and check whether backend should not work without token

Srinivas-Anand-HPE replied:
It's  output is compared and handled below

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+                result = getattr(requests, method_func)(url, headers=headers, json=payload, verify=False, timeout=timeout)
+            else:
+                result = getattr(requests, method_func)(url, headers=headers, verify=False, timeout=timeout)
+        except:
+            raise
```
If we are just immediately re-raising the exception and doing nothing, why is this in a try block at all? The exception will be raised on its own if there is no try block here.

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+        try:
+            self.get_activity(activity)
+            return True
+        except:
```
Bare `except` statements should be avoided. They catch everything, including fatal system-level exceptions, which is not usually a good thing to do.

Ideally, all except statements should specify the types of exceptions they are going to catch. But in cases where that is not feasible, it should at least just catch `Exception`. That will not blindly catch super-low-level exceptions.

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,257 @@
+            self.get_activity(activity)
+            return True
+        except:
+            return False
```
Do we really not want to do any kind of logging about the exception that was raised? This acts the same way if the request was a "404 not found" as it would if it was a "500 service unavailable", or if there was some other non-request-related Python exception that got raised. It seems like this function should return False only in cases where the activity actually doesn't exist. Wouldn't it make more sense for it to raise the exception in the case that some unexpected error happened?

mharding-hpe replied:
[This comment from the previous review](https://github.com/Cray-HPE/csm-testing/pull/629/files#r1873813589) has not been addressed or responded to

Srinivas-Anand-HPE replied:
@mharding-hpe  

if not apis.activity_exists(activity) i.e if it returs in the scenario where it does not successfully return that activity in existing or backend is not working as excepted 
msg = f"ERROR: Activity {activity} does not exist. or Backend is not working properly"
print_no_of_test_passed(msg, count) . This where we print summary of number of test cases passed and and number of test cases pending and exit in the function with an exit code 1

Srinivas-Anand-HPE replied:
`
def print_no_of_test_passed(ex, count: int):
    print(f"ERROR: {ex}")
    print("~" * 50)
    print(f"INFO: TOTAL test cases passed: {count} test cases skipped: {15 - count}")
    print("~" * 50)
    sys.exit(1)`

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,259 @@
+from kubernetes.client.rest import ApiException
+import yaml
+
+def is_main_container_finished(pod_name, namespace):
```
```suggestion
def is_main_container_finished(pod_name: str, namespace: str) -> bool:
```

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,259 @@
+            v1.delete_namespaced_pod(name=pod_name, namespace=namespace)
+            print(f"INFO: Pod {pod_name} deleted.")
+            return True
+        else:
+            print(f"INFO: Main container still running in pod {pod_name}.")
```
No need for this to be in an `else` block, since the first branch ended with a `return` statement.
```suggestion
        print(f"INFO: Main container still running in pod {pod_name}.")
```

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,259 @@
+            print(f"INFO: Main container finished in pod {pod_name}. Deleting pod.")
+            v1.delete_namespaced_pod(name=pod_name, namespace=namespace)
+            print(f"INFO: Pod {pod_name} deleted.")
+            return True
```
If this function returns `True` in this case, should it return `False` at the end of the function, in the case where no delete happened?

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,259 @@
+                    print(f"WARNING: Workflow {workflow_name} is running but main container is done.")
+                    return True
+        except ApiException as e:
+            print(f"ERROR: Exception when fetching workflow status: {e}")
```
That message makes it look like the try statement should be around the `kubectl` command, since that is what is retrieving the status.

But this `except` block is catching `ApiException`, which seems like it would only be raised by the call to `check_and_kill_pod`. 

Either way, the try block should not enclose such a large block of code. It should surround the part that is expected to raise the exception, it should catch the appropriate exception, and the message should be relevant to what was being attempted.

Pankhuri-Rajesh replied:
I moved the except block , now it will catch subprocess error for kubectl command.
```
try:
            # Get the workflow status
            command = [
                "kubectl",
                "get",
                "workflow",
                workflow_name,
                "-n",
                namespace,
                "-o",
                "jsonpath={.status.phase}",
            ]
            result = subprocess.run(
                command,
                check=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                universal_newlines=True,
            )
        except subprocess.CalledProcessError as err:
            print(f"ERROR: Exception when fetching workflow status: {err}")
            return False
```

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,259 @@
+        if not pod_names:
+            print(f"INFO: No pods found for workflow {workflow_name}.")
+    except subprocess.CalledProcessError as e:
+        print(f"ERROR: Failed to delete resources: {e}")
```
The exception here could have been raised by the command to list the pods, but the message here seems to assume that it arose from the delete command. Here again, the try blocks should surround just the applicable code, and provide a response that is appropriate for that code.

mharding-hpe commented on src/csm_testing/tests/iuf_stages_cleanup/__main__.py:
```diff
@@ -0,0 +1,139 @@
+def cleanup_deliver_product():
+    repo_name = "dummy-repo"
+    username,password = get_nexus_credentials()
+    """Delete repository from Nexus."""
```
If this is meant to be a docstring for the function, it should appear immediately after the `def` line.
If it is meant to be a comment, it should begin with a `#`

mharding-hpe commented on src/csm_testing/tests/sat_validate_bootprep/__main__.py:
```diff
@@ -0,0 +1,73 @@
+
+class BootPrepValidationError(Exception):
+    """Custom exception for BootPrep validation errors."""
+    pass
```
The `pass` is not needed since the class has a docstring.

```suggestion
```

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,416 @@
+                check_ims_cmd=f"cray ims images describe {final_image_id.strip()}"
+                ims_info, _ = run_command(check_ims_cmd)
+                if not ims_info:
+                        print(f"ERROR: Could not find ims image for {name}")
+                        print(f"INFO: Total test cases executed for stage operations: {test_cases}")
+                        sys.exit(1)
```
These lines are indented too far. It won't cause a syntax error, but ideally the indenting convention should remain consistent.

```suggestion
               print(f"ERROR: Could not find ims image for {name}")
               print(f"INFO: Total test cases executed for stage operations: {test_cases}")
               sys.exit(1)
```

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,416 @@
+# are helper functions being used by stage operations
+    
+def nexus_repo_setup():
+    global test_cases
```
This is unnecessary, since the variable isn't being assigned to in this function.

```suggestion
```

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,416 @@
+        return False
+    return True
+
+def run_iuf_script(*args):
```
The `args` argument is not used in the function. Why is it here?

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,416 @@
+        return False
+    return True
+
+def process_media(*args):
```
The `args` argument is not used in the function. Why is it here?

mharding-hpe commented on src/csm_testing/tests/iuf_apis/__main__.py:
```diff
@@ -0,0 +1,114 @@
+"""
+
+
+import requests
+import os
+from kubernetes import client, config
+import base64
+from keycloak import KeycloakOpenID
```
None of these imports are used in this file, that I can see.

```suggestion
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/__main__.py:
```diff
@@ -0,0 +1,114 @@
+import urllib3
+import time
+import sys
+import copy
```
Unused import.
```suggestion
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/__main__.py:
```diff
@@ -0,0 +1,114 @@
+from csm_testing.lib.iuf_constants import MEDIA_DIR
+from csm_testing.lib.iuf_common import media_dir_setup
+
+from csm_testing.lib.iuf_classes import AuthException
+from csm_testing.lib.iuf_classes import Auth 
```
Unused imports.

```suggestion
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/apis_session_function.py:
```diff
@@ -0,0 +1,188 @@
+from .apis_activity_functions import count, total_count
+
+def get_sessions(apis, activity):
+    global count , total_count
```
`total_count` is not assigned to in this function, so it does not need to be declared here. 
```suggestion
    global count
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/apis_session_function.py:
```diff
@@ -0,0 +1,188 @@
+        sys.exit(1)
+
+def get_workflows(apis, activity):
+    global count , total_count
```
```suggestion
    global count
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/apis_session_function.py:
```diff
@@ -0,0 +1,188 @@
+        sys.exit(1)
+
+def get_history(apis, activity):
+    global count , total_count
```
```suggestion
    global count
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/apis_session_function.py:
```diff
@@ -0,0 +1,188 @@
+        sys.exit(1)
+
+def get_history_time(apis, activity):
+    global count , total_count
```
```suggestion
    global count
```

mharding-hpe commented on src/csm_testing/tests/iuf_apis/apis_session_function.py:
```diff
@@ -0,0 +1,188 @@
+        sys.exit(1)
+
+def get_activity_session(apis, activity):
+    global count , total_count
```
```suggestion
    global count
```

mharding-hpe reviewed (changes requested):
Most of these issues are general Python things, not specific to `csm-testing`. I suggest making sure this code is peer-reviewed on your end, specifically to look for issues in the Python code. I'd also suggest looking at the pylint results in the build pipeline. The non-fatal pylint issues won't cause the build to fail, but they are still identified in the log, and they can point out potential problems or bad practices.

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+    """A wrapper for raising an AuthException exception."""
+
+
+class Auth:  # pylint: disable=missing-class-docstring
```
We do not want to resolve the pylint errors just by adding these disable statements. The disable statements are intended to be used for cases like:
1. pylint is complaining about things that are not really errors. The best example I can think of for this is the test which imports the `rados` package. That package is not installed when we run pylint, so we get an import error, but we know that this package will be available on the actual systems. This is because the RPM itself requires that this is installed in the system Python environment.
2. pylint is complaining about something that is being done deliberately and we have a good reason not to change it.
3. There are a small number of cases where the pylint complaint is valid, but we know we will never address it. The only example of this I can think of is that we generally add `disable=missing-function-docstring` for the `main()` methods in the `__main__.py` files. This is because that function is the main body of the test or tool, and the module-level docstring should already contain a description of the test or tool logic. So repeating it in the `main()` function docstring would mostly be redundant. I suppose this could also be the case for extremely small, simple, self-explanatory functions.

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+import base64
+from urllib.error import HTTPError
+import requests
+from kubernetes import client, config  # pylint: disable=import-error
+from keycloak import KeycloakOpenID  # pylint: disable=import-error
```
Neither of these should be raising import errors when pylint runs during the builds. Why are the pylint disable comments added?

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+    def __init__(self):  # pylint: disable=missing-function-docstring
+        self._token = None
+
+    def get_secrets(self):  # pylint: disable=missing-function-docstring,R0201
```
Per the other comment, we should be minimizing pylint disable statements. If we DO need to include them, we should not specify the things being ignored using their codes like `R0201`. We should always use the descriptive name (like `no-self-use`, in the case of `R0201`)

mharding-hpe replied:
This is an example where it is just as easy to address the pylint error than it is to ignore it. The `no-self-use` error is saying that it doesn't make sense for this to be a regular method of the class, since it does not make use of the `self` argument. Thus, it makes more sense for it to be either a regular (non-class) function, or a static method, or a class method. This can be done easily. 

Also, as mentioned in a previous comment, we should include type annotations as much as possible.

For example:

Add to the imports earlier:
```python
from typing import Tuple
```

```suggestion
    @classmethod
    def get_secrets(cls) -> Tuple[str, str]:
```

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+    def get_secrets(self):  # pylint: disable=missing-function-docstring,R0201
+        try:
+            config.load_kube_config()
+            v1 = client.CoreV1Api()  # pylint: disable=invalid-name
+            sec = v1.read_namespaced_secret("admin-client-auth", "default").data
```
While this is not a serious problem being reported by pylint, it is another example where it is just as easy to fix the problem as it is to ignore the error.
```suggestion
            k8s_v1 = client.CoreV1Api()
            sec = k8s_.read_namespaced_secret("admin-client-auth", "default").data
```

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+            password = base64.b64decode(sec.get("client-secret").strip()).decode(
+                "utf-8"
+            )
+        except:  # pylint: disable=raise-missing-from
```
As [noted previously](https://github.com/Cray-HPE/csm-testing/pull/629#discussion_r1873810902), we really do not want bare exceptions like this. In the worst case, we can do `except Exception`. It is far better to catch more specific exceptions than that, of course.

Is there something we need to catch here that is not covered by `Exception`?

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+
+
+class Auth:  # pylint: disable=missing-class-docstring
+    def __init__(self):  # pylint: disable=missing-function-docstring
```
I don't think pylint complains when `__init__` methods lack docstrings (I assume because their content would usually be covered in the class docstring itself). We want to avoid adding disable statements for warnings that pylint is not actually raising.

```suggestion
    def __init__(self):
```

mharding-hpe commented on src/csm_testing/tests/iuf_cleanup/__main__.py:
```diff
@@ -0,0 +1,220 @@
+        return err.returncode
+    return configmaps
+
+# pylint: disable=too-many-branches
```
If pylint is raising this error, it is a strong indication that the function in question should be refactored, possibly splitting it into multiple functions. This is one of the more important pylint errors to actually address, because it can make code much harder to understand and maintain.

Pankhuri-Rajesh replied:
we split the function into two. resolved the warning. we are catching general Exception and even if the code is not able to delete anything , it should just continue and try to delete the remaining directories/cms/wfls.

mharding-hpe commented on src/csm_testing/tests/iuf_stages/__main__.py:
```diff
@@ -0,0 +1,484 @@
+        print(f"INFO: Total test cases executed for stage operations: {TEST_CASES}")
+        sys.exit(1)
+
+def prepare_images(activity_name): # pylint: disable=too-many-locals
```
If pylint is raising the `too-many-locals` error, it is a strong indication that the function in question should be refactored, possibly splitting it into multiple functions. This is one of the more important pylint errors to actually address, because it can make code much harder to understand and maintain.

Pankhuri-Rajesh replied:
I split the function in 2 parts. Also I split iuf_stages into 2 files - __main__.py and helper.py
helper.py will have all the helper functions.

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+
+        return username, password
+
+    def get_token(
+        self, username, password
+    ):  # pylint: disable=missing-function-docstring,R0201
```
```suggestion
    @classmethod
    def get_token(
        cls, username: str, password: str
    ) -> str:
```

mharding-hpe commented on src/csm_testing/lib/iuf_classes.py:
```diff
@@ -0,0 +1,277 @@
+        return token["access_token"]
+
+    @property
+    def token(self):  # pylint: disable=missing-function-docstring
```
```suggestion
    def token(self):
```

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,314 @@
+    return False
+
+
+def check_and_kill_pod(workflow_name, namespace):
```
```suggestion
def check_and_kill_pod(workflow_name: str, namespace: str) -> bool:
```

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,314 @@
+
+
+def check_and_kill_pod(workflow_name, namespace):
+    """Kills the pod if the main container is finished"""
```
The docstring should explain what the meaning of the return value is.

mharding-hpe commented on goss-testing/vars/variables-ncn.yaml:
```diff
@@ -163,3 +163,32 @@ storage_fs_labels:
 vshasta: false
+
+
+pre_checks:
```
The variables you are adding should probably all be prefaced with `iuf_`, except possibly the `bootprep` one. Otherwise it is not very obvious to someone looking at this file what they are related to.

Pankhuri-Rajesh replied:
except bootprep we renamed all variables

mharding-hpe commented on constraints.txt:
`packaging`  should be added to this file, so we don't just take the latest one every time. Same for any dependencies that are pulled in by any of the new packages being added into `requirements.txt`, if they're not already present in here. The build pipeline output shows a snapshot of what gets installed into the virtual environment, so you can use that as the basis.

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,314 @@
+                    )
+                    return True
+        except ApiException as err:
+            print(f"ERROR: Exception when fetching workflow status: {err}")
```
[This comment from a previous review](https://github.com/Cray-HPE/csm-testing/pull/629#discussion_r1873822841) was not addressed.

Pankhuri-Rajesh replied:
I have modified the code to wrap the kubectl command in try-Exception block.

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,314 @@
+            print(f"ERROR: Failed to delete {resource} {name}: {err}")
+            continue
+
+    try:
+        print(f"INFO: Finding pods owned by workflow {workflow_name}...")
+        pods_json = subprocess.run(
+            ["kubectl", "get", "pods", "-n", namespace, "-o", "json"],
+            check=True,
+            stdout=subprocess.PIPE,
+            universal_newlines=True,
+        ).stdout
+        pod_names = [
+            pod["metadata"]["name"]
+            for pod in json.loads(pods_json).get("items", [])
+            if any(
+                owner.get("name") == workflow_name
+                for owner in pod.get("metadata", {}).get("ownerReferences", [])
+            )
+        ]
+    except subprocess.CalledProcessError as err:
+        print(f"ERROR: Failed to retrieve pod list for workflow {workflow_name}: {err}")
+        return
+    except json.JSONDecodeError as err:
+        print(f"ERROR: Failed to parse pod list: {err}")
+        return
```
```suggestion
    try:
        print(f"INFO: Finding pods owned by workflow {workflow_name}...")
        pods_json = subprocess.run(
            ["kubectl", "get", "pods", "-n", namespace, "-o", "json"],
            check=True,
            stdout=subprocess.PIPE,
            universal_newlines=True,
        ).stdout
    except subprocess.CalledProcessError as err:
        print(f"ERROR: Failed to retrieve pod list for workflow {workflow_name}: {err}")
        return
    try:
        pod_names = [
            pod["metadata"]["name"]
            for pod in json.loads(pods_json).get("items", [])
            if any(
                owner.get("name") == workflow_name
                for owner in pod.get("metadata", {}).get("ownerReferences", [])
            )
        ]
    except json.JSONDecodeError as err:
        print(f"ERROR: Failed to parse pod list: {err}")
        return
```

mharding-hpe commented on src/csm_testing/tests/check_workflow_template/__main__.py:
```diff
@@ -0,0 +1,314 @@
+            )
+            print(f"INFO: {resource.capitalize()} {name} deleted successfully.")
+        except subprocess.CalledProcessError as err:
+            print(f"ERROR: Failed to delete {resource} {name}: {err}")
```
Various print statements in this function indicate errors, but no exception is raised, and the function returns no value to indicate that errors happened. Should the test not fail in the case where these operations fail?

Pankhuri-Rajesh replied:
this is intentional. Even if one resource is not deleted, we want to continue and try for other resources and just to leave a trace of failure we have added Exceptions

mharding-hpe replied:
But it seems like the test will still pass, in that case, even though there was a problem. You can still have the cleanup cover the other resources and then have the test end up failing. For example, you could modify the Goss YAML test files so that they fail if they see 'ERROR' in the output of the test.

Pankhuri-Rajesh replied:
we don't want to test case to fail in case the cleanup is not successful. So rather than keeping the log as ERROR statements, I have made them as WARNING.

mharding-hpe replied:
This is fine as long as the resources left behind aren't things that will cause problems for anyone else using the system (or cause future iterations of the test to fail).

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-abort.yaml:
```diff
@@ -0,0 +1,62 @@
+        "{{$iuf_cleanup}}" "{{ index $activity 0 }}"
+
+        # Exit with code 1 if the iuf abort command failed with exit code 1
+        if [ $run_exit_code -eq 1 ]; then
```
Shouldn't the test fail if iuf_abort returned any non-0 exit code? Or are some non-0 returns expected and not a problem? Same comment for the other tests.

mharding-hpe commented on src/csm_testing/tests/iuf_pre_checks/__main__.py:
```diff
@@ -0,0 +1,535 @@
+from csm_testing.lib.iuf_common import run_command
+
+
+# compare_versions is a helper functions being used by other functions
```
```suggestion
```

mharding-hpe commented on src/csm_testing/tests/iuf_pre_checks/__main__.py:
```diff
@@ -0,0 +1,535 @@
+            sys.exit(1)
+
+
+# function to check ssh connectivity
```
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-abort.yaml:
```diff
@@ -0,0 +1,62 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
+{{ $iuf_run := $scripts | printf "%s/python/iuf_run" }}
```
Remove unused test variables.
```suggestion
{{ $activity := .Vars.iuf_activity }}
{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-apis.yaml:
```diff
@@ -0,0 +1,62 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
+{{ $iuf_run := $scripts | printf "%s/python/iuf_run" }}
```
Remove unused test variables.
```suggestion
{{ $activity := .Vars.iuf_activity }}
{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-negative.yaml:
```diff
@@ -0,0 +1,67 @@
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-restart.yaml:
```diff
@@ -0,0 +1,65 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-resume.yaml:
```diff
@@ -0,0 +1,63 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-activity-abort.yaml:
```diff
@@ -0,0 +1,62 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
+{{ $iuf_run := $scripts | printf "%s/python/iuf_run" }}
```
Remove unused test variables.
```suggestion
{{ $activity := .Vars.iuf_activity }}
{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-activity-abort.yaml:
```diff
@@ -0,0 +1,62 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
+{{ $iuf_run := $scripts | printf "%s/python/iuf_run" }}
+{{ $iuf_abort := $scripts | printf "%s/check_iuf_abort.sh" }}
```
Remove unused test variables.
```suggestion
{{ $activity := .Vars.iuf_activity }}
{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-activity-resume.yaml:
```diff
@@ -0,0 +1,60 @@
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := "/etc/cray/upgrade/csm/dummy" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-restart.yaml:
```diff
@@ -0,0 +1,65 @@
+{{ $activity := .Vars.iuf_activity }}
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
+{{ $iuf_run := $scripts | printf "%s/python/iuf_run" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-restart.yaml:
```diff
@@ -0,0 +1,65 @@
+          # Cleanup the workflows,log and media dir
+          "{{$iuf_cleanup}}"
+
+          # Exit with code 1 if the iuf_run command failed with exit code 1
```
```suggestion
          # Exit with code 1 if the iuf command failed with exit code 1
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-resume.yaml:
```diff
@@ -0,0 +1,63 @@
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
+{{ $logrun := .Env.GOSS_BASE | printf "%s/scripts/log_run.sh" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
+{{ $iuf_run := $scripts | printf "%s/python/iuf_run" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-resume.yaml:
```diff
@@ -0,0 +1,63 @@
+          # Cleanup the workflows,log and media dir
+          "{{$iuf_cleanup}}"
+
+          # Exit with code 1 if the iuf_run command failed with exit code 1
```
```suggestion
          # Exit with code 1 if the iuf command failed with exit code 1
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-media.yaml:
```diff
@@ -0,0 +1,60 @@
+        # Run cleanup step
+        "{{$iuf_cleanup}}" "{{ index $activity 2 }}"
+
+        # Exit with code 1 if the iuf_run command failed
```
```suggestion
        # Exit with code 1 if the iuf command failed
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-list-stages-negative.yaml:
```diff
@@ -0,0 +1,76 @@
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $activity := .Vars.iuf_activity }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $media_dir := .Env.GOSS_BASE | printf "%s/scripts/iuf_run_setup" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-no-media.yaml:
```diff
@@ -0,0 +1,58 @@
+        # Run cleanup step
+        "{{$iuf_cleanup}}" "{{ index $activity 2 }}"
+
+        # Exit with code 1 if the iuf_run command failed
```
```suggestion
        # Exit with code 1 if the iuf command failed
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-validate-iuf-manifest.yaml:
```diff
@@ -0,0 +1,45 @@
+
+{{ $pre_checks := .Vars.iuf_pre_checks }}
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $iuf_pre_checks := $scripts | printf "%s/python/iuf_pre_checks" }}
```
Remove unused test variables.
```suggestion
```

mharding-hpe commented on goss-testing/vars/variables-ncn.yaml:
```diff
@@ -163,3 +163,32 @@ storage_fs_labels:
 pit_node: false
 # false by default--overriden when vshasta is deployed
 vshasta: false
+
```
```suggestion
```

mharding-hpe commented on src/csm_testing/lib/iuf_common.py:
```diff
@@ -0,0 +1,196 @@
+    return dummy_repo_url, auth, ca_cert_path
+
+
+# Custom exception for IUF Product Manifest validation errors
```
```suggestion
```

mharding-hpe commented on src/csm_testing/tests/validate_iuf_manifest/__main__.py:
```diff
@@ -0,0 +1,91 @@
+SCHEMA_FILE = "/opt/cray/tests/install/ncn/scripts/iuf_schemas/iuf-manifest-schema.yaml"
+
+
+# Custom exception for IUF Product Manifest validation errors
```
```suggestion
```

haasken-hpe commented:
Hi @Pankhuri-Rajesh, can we please set up a time to discuss the intention of these tests and the possible overlap with the automated tests that the SAT team (me, @ethanholen-hpe, and @annapoorna-s-alt) is developing?

One thing I saw here really concerns me, which is that the entire JSON Schema files (for bootprep input files and IUF product manifest files) are being copied into this repository. I would like to understand why that is being done and come up with a better solution.

Thank you!

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-abort.yaml:
```diff
@@ -0,0 +1,60 @@
+        "{{$iuf_cleanup}}" "{{ index $activity 0 }}"
+
+        # Exit with non-zero exit code if the iuf abort command failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-apis.yaml:
```diff
@@ -0,0 +1,60 @@
+        "{{$iuf_cleanup}}" "{{ index $activity 0 }}"
+
+        # Exit with non-zero exit code if the iuf apis script failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-list-stages.yaml:
```diff
@@ -0,0 +1,64 @@
+          "{{$iuf_cleanup}}" "{{ index $activity 0 }}"
+
+          # Exit with non-zero exit code if the iuf ls command failed with non-zero exit code
+          if [ $run_exit_code -ne 0 ]; then
+            exit $run_exit_code
+          fi
```
```suggestion
          exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-negative.yaml:
```diff
@@ -0,0 +1,66 @@
+          "{{$iuf_cleanup}}" "{{ index $activity 2 }}"
+
+          # Exit with non-zero exit code if the iuf activity command failed with non-zero exit code
+          if [ $run_exit_code -ne 0 ]; then
+            exit $run_exit_code
+          fi
```
```suggestion
          exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-restart.yaml:
```diff
@@ -0,0 +1,63 @@
+          "{{$iuf_cleanup}}"
+
+          # Exit with non-zero exit code if the iuf command failed with exit non-zero exit code
+          if [ $run_exit_code -ne 0 ]; then
+            exit $run_exit_code
+          fi
```
```suggestion
          exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity-resume.yaml:
```diff
@@ -0,0 +1,61 @@
+          "{{$iuf_cleanup}}"
+
+          # Exit with non-zero exit code if the iuf command failed with non-zero exit code
+          if [ $run_exit_code -ne 0 ]; then
+            exit $run_exit_code
+          fi
```
```suggestion
          exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-activity.yaml:
```diff
@@ -0,0 +1,68 @@
+        "{{$iuf_cleanup}}" "{{ index $activity 0 }}"
+
+        # Exit with non-zero exit code if the iuf activity failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-activity-abort.yaml:
```diff
@@ -0,0 +1,59 @@
+        # Exit with non-zero exit code if the iuf abort command failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-activity-restart.yaml:
```diff
@@ -0,0 +1,59 @@
+          "{{$iuf_cleanup}}" "{{ index $activity 2 }}"
+
+          # Exit with non-zero exit code if the iuf restart command failed with non-zero exit code
+          if [ $run_exit_code -ne 0 ]; then
+            exit $run_exit_code
+          fi
```
```suggestion
          exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-invalid-activity-resume.yaml:
```diff
@@ -0,0 +1,59 @@
+      "{{$iuf_cleanup}}" "{{ index $activity 2 }}"
+
+      # Exit with non-zero exit code if the iuf resume command failed with non-zero exit code
+      if [ $run_exit_code -ne 0 ]; then
+        exit $run_exit_code
+      fi
```
```suggestion
      exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-list-activity-workflows.yaml:
```diff
@@ -0,0 +1,64 @@
+        "{{$iuf_cleanup}}"
+
+        # Exit with non-zero exit code if the iuf -a <activity> workflow command failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-list-stages-negative.yaml:
```diff
@@ -0,0 +1,75 @@
+        "{{$iuf_cleanup}}" "{{ index $activity 2 }}"
+
+        # Exit with non-zero exit code if the iuf -a <activity> ls command failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-run.yaml:
```diff
@@ -0,0 +1,61 @@
+        "{{$iuf_cleanup}}" "{{ index $activity 0 }}"
+
+        # Exit with non-zero exit code if the iuf_run command failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-specified-log-dir.yaml:
```diff
@@ -0,0 +1,68 @@
+        rm -r "{{ index $logdir 1}}"
+
+        # Exit with non-zero exit code if the iuf_run command failed with non-zero exit code
+        if [ $run_exit_code -ne 0 ]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-stages.yaml:
```diff
@@ -0,0 +1,64 @@
+        # Exit with non-zero exit code if the stage operations script failed with non-zero exit code
+        if [ $run_exit_code -ne 0]; then
+          exit $run_exit_code
+        fi
```
```suggestion
        exit $run_exit_code
```

mharding-hpe commented on goss-testing/tests/ncn/goss-sat-validate-management-bootprep.yaml:
```diff
@@ -0,0 +1,43 @@
+      sev: 0
+    exec: |-
+      # validate management-bootprep
+      
```
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-validate-iuf-manifest.yaml:
```diff
@@ -0,0 +1,44 @@
+      desc: Validates iuf_product_manifest.yaml
+      sev: 0
+    exec: |-
+
```
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-sat-validate-compute-uan-bootprep.yaml:
```diff
@@ -0,0 +1,43 @@
+      sev: 0
+    exec: |-
+      # validate compute-and-uan-bootprep
+      
```
```suggestion
```

mharding-hpe commented on goss-testing/tests/ncn/goss-iuf-workflowtemplates.yaml:
```diff
@@ -0,0 +1,75 @@
+        # Exit with code 1 if test case failed with exit code 1
+        if [ $run_exit_code_1 -eq 1 ] || [ $run_exit_code_2 -eq 1 ]; then
+          exit 1
+        fi
```
Should this also not fail if the subtests returned other non-0 return codes?
```suggestion
        # Fail if either subtest failed
        [[ $run_exit_code_1 -ne 0 ]] && exit $run_exit_code_1
        exit $run_exit_code_2
```

mharding-hpe commented on src/csm_testing/tests/iuf_post_checks/__main__.py:
```diff
@@ -0,0 +1,155 @@
+    except subprocess.CalledProcessError as err:
+        print(
+            f"ERROR: ConfigMap '{configmap_name}' does NOT exist in the '{namespace}' namespace.\n\
+Error: {err}")
```
```suggestion
            f"ERROR: ConfigMap '{configmap_name}' does NOT exist in the '{namespace}' namespace.\n"
            f"Error: {err}")
```

mharding-hpe commented on src/csm_testing/tests/iuf_cleanup/__main__.py:
```diff
@@ -0,0 +1,212 @@
+        activity_name(str): The activity name whose configmaps list needs to be generated
+
+    """
+    command = f"kubectl get configmap -n argo -o \
+custom-columns=NAME:.metadata.name|grep {activity_name}"
```
```suggestion
    command = ("kubectl get configmap -n argo -o "
               f"custom-columns=NAME:.metadata.name|grep {activity_name}")
```

mharding-hpe commented on src/csm_testing/tests/iuf_cleanup/__main__.py:
```diff
@@ -0,0 +1,212 @@
+    Args:
+        activity_name(str): The activity name whose workflows list needs to be generated
+    """
+    command = f"kubectl get workflow -n argo -o \
+custom-columns=NAME:.metadata.name|grep {activity_name}"
```
```suggestion
    command = ("kubectl get workflow -n argo -o "
               f"custom-columns=NAME:.metadata.name|grep {activity_name}")
```

mharding-hpe commented on src/csm_testing/tests/validate_iuf_manifest/__main__.py:
```diff
@@ -0,0 +1,88 @@
+    # Validate the product manifest file against the schema
+    try:
+        validate_instance(manifest_instance, schema)
+        print(f"INFO: SUCCESS: IUF product manifest file \
+'{manifest_file}' is valid against the schema.")
```
```suggestion
        print("INFO: SUCCESS: IUF product manifest file "
              f"'{manifest_file}' is valid against the schema.")
```

mharding-hpe reviewed (changes requested):
My feedback has been addressed. I am marking "request changes" for my review because I don't want to merge until @haasken-hpe  approves. It sounds like he has some concerns, [based on his comemnt](https://github.com/Cray-HPE/csm-testing/pull/629#issuecomment-2540162573)

haasken-hpe commented:
Thanks, @mharding-hpe. @Pankhuri-Rajesh set up a meeting with me and the SAT team today, and we talked about my feedback from my earlier comment.

With respect to the inclusion of the `sat bootprep` input file schema copied from [bootprep_schema.yaml in the sat repo](https://github.com/Cray-HPE/sat/blob/main/sat/data/schema/bootprep_schema.yaml), I informed the IUF team that it is possible to obtain this schema file directly from the `sat` command by running `sat bootprep view-schema`. Obtaining it in this way has a couple benefits.

* We will not duplicate the `sat bootprep` schema file in this repo.
* The version obtained from `sat bootprep view-schema` will be the correct version for the version of the `sat` command on the system.

I also mentioned that `sat bootprep run` has a `--dry-run` option that could be useful for performing schema validation. I did point out that this command will do more than just schema validation, and it will require that the variables used in the bootprep file are defined, products referenced in the bootprep file are present in the product catalog, and requested branches of VCS repos exist. If none of that is a problem for this test case, then the `sat bootprep run --dry-run` command could be used.

In the meeting between the SAT and IUF teams, we also discussed the possible overlap between the test cases being added by the IUF team and those added by the SAT team. I don't believe there will be significant overlap, and we plan to include the IUF team members on reviews for the SAT tests moving forward to ensure awareness of the tests we are adding as well.

We did not talk about the concern I have about the IUF product manifest schema being duplicated in this repository. That's not really my area anymore, but I can say that the `cray-nls` image does provide a `validate` command that can be used to validate the schema of a given IUF product manifest file against the schema contained in that `cray-nls` image. I would suggest that this should be used instead, if possible. I believe that most products are doing this validation already in the build of their product's release distribution tar file. See [release-csm.sh in the USS product stream](https://github.hpe.com/hpe/hpc-shasta-os-uss-product-stream/blob/master/release-csm.sh#L345), for example. The function `iuf-validate` is defined in [the release.sh library script in hpc-shastarelm-release](https://github.hpe.com/hpe/hpc-shastarelm-release/blob/master/lib/release.sh#L895). Perhaps these tests can follow that example to do something similar.

haasken-hpe commented:
@Pankhuri-Rajesh, please let me know when the feedback about the SAT bootprep schema file has been addressed, and I will take a look at just that piece and approve the PR, so it's clear my feedback has been addressed. I can also take a look at the IUF schema validation when that's addressed, but I'm not the expert there.

Pankhuri-Rajesh commented:
> Thanks, @mharding-hpe. @Pankhuri-Rajesh set up a meeting with me and the SAT team today, and we talked about my feedback from my earlier comment.
> 
> With respect to the inclusion of the `sat bootprep` input file schema copied from [bootprep_schema.yaml in the sat repo](https://github.com/Cray-HPE/sat/blob/main/sat/data/schema/bootprep_schema.yaml), I informed the IUF team that it is possible to obtain this schema file directly from the `sat` command by running `sat bootprep view-schema`. Obtaining it in this way has a couple benefits.
> 
> * We will not duplicate the `sat bootprep` schema file in this repo.
> * The version obtained from `sat bootprep view-schema` will be the correct version for the version of the `sat` command on the system.
> 
> I also mentioned that `sat bootprep run` has a `--dry-run` option that could be useful for performing schema validation. I did point out that this command will do more than just schema validation, and it will require that the variables used in the bootprep file are defined, products referenced in the bootprep file are present in the product catalog, and requested branches of VCS repos exist. If none of that is a problem for this test case, then the `sat bootprep run --dry-run` command could be used.
> 
> In the meeting between the SAT and IUF teams, we also discussed the possible overlap between the test cases being added by the IUF team and those added by the SAT team. I don't believe there will be significant overlap, and we plan to include the IUF team members on reviews for the SAT tests moving forward to ensure awareness of the tests we are adding as well.
> 
> We did not talk about the concern I have about the IUF product manifest schema being duplicated in this repository. That's not really my area anymore, but I can say that the `cray-nls` image does provide a `validate` command that can be used to validate the schema of a given IUF product manifest file against the schema contained in that `cray-nls` image. I would suggest that this should be used instead, if possible. I believe that most products are doing this validation already in the build of their product's release distribution tar file. See [release-csm.sh in the USS product stream](https://github.hpe.com/hpe/hpc-shasta-os-uss-product-stream/blob/master/release-csm.sh#L345), for example. The function `iuf-validate` is defined in [the release.sh library script in hpc-shastarelm-release](https://github.hpe.com/hpe/hpc-shastarelm-release/blob/master/lib/release.sh#L895). Perhaps these tests can follow that example to do something similar.

@haasken-hpe  I have made the following changes after our discussion-

1. Instead of placing schema in the repo, "sat bootprep view-schema" is being used to validate the manifest files.
2. For IUF manifest validation we have replaced the previous method with the podman command which uses cray-nls:0.10.0 image to do the validation. ( referenced from [the release.sh library script in hpc-shastarelm-release](https://github.hpe.com/hpe/hpc-shastarelm-release/blob/master/lib/release.sh#L895) )
3. I have also removed the commented lines from management-bootprep file as discussed.
4. Hence, iuf_schemas folder has been deleted .
5. I have tested the changed code on noname and it works

Please review the PR changes, and if it's ok then you can approve. Let me know if you have any further follow up on this.

haasken-hpe commented on src/csm_testing/tests/validate_iuf_manifest/__main__.py:
```diff
@@ -0,0 +1,99 @@
+
+    try:
+        cray_nls_image = (
+            "arti.hpc.amslabs.hpecorp.net/csm-docker-remote/stable/cray-nls:0.10.0"
```
Is this path and version of the image always going to be available where these tests are executed?

haasken-hpe reviewed (approved):
The bootprep schema validation change looks good, and I'm glad to see the schema files removed from this PR.

spillerc-hpe commented on src/csm_testing/tests/iuf_pre_checks/__main__.py:
```diff
@@ -0,0 +1,554 @@
+        f"https://argo.cmn.{cluster_name}.hpc.amslabs.hpecorp.net/",
+        f"https://vcs.cmn.{cluster_name}.hpc.amslabs.hpecorp.net/",
+        f"https://nexus.cmn.{cluster_name}.hpc.amslabs.hpecorp.net/",
+    ]
```
hpc.amslabs.hpecorp.net is an internal domain, no customer system will have these values.

One way would be to read the SYSTEM_DOMAIN environment variable if it's available here. Another would be to interrogate Kubernetes for the list of virtual services (`kubectl get virtualservice -A`)

Pankhuri-Rajesh replied:
@spillerc-hpe  I have made the change in the script so that we get the urls using "kubectl get virtualservice" command. can you review and let us know if it is fine

spillerc-hpe replied:
lgtm

mharding-hpe commented:
I see that there is [an outstanding comment](https://github.com/Cray-HPE/csm-testing/pull/629#discussion_r1892571355) by @spillerc-hpe. I'm holding off on merging until Chris has approved the PR. I've asked @Srinivas-Anand-HPE to follow up with Chris and make sure his concerns are addressed.

mharding-hpe commented:
/backport release/1.7
